scripts/subspace_mlp_demo.py

```diff
@@ -0,0 +1,191 @@
+# This demo replicates Figure 2 of the paper
+# "Measuring the Intrinsic Dimension of Objetive Landscape"
+# By Li et al. (https://arxiv.org/abs/1804.08838)
+# We consider a 2-layer MLP with ReLU activations
+# Code based on the following repos:
+# * https://github.com/ganguli-lab/degrees-of-freedom
+# * https://github.com/uber-research/intrinsic-dimension
+
+# Author: Gerardo Durán-Martín (@gerdm)
+
+import superimport
+
+import jax
+import optax
+import jax.numpy as jnp
+import flax.linen as nn
+import numpy as np
+import tensorflow_datasets as tfds
+import matplotlib.pyplot as plt
+import pyprobml_utils as pml
+from jax import random
+from time import time
+from functools import partial
+
+
+def get_datasets():
+    """
+    Load MNIST train and test datasets into memory
+    """
+    ds_builder = tfds.builder("mnist")
+    ds_builder.download_and_prepare()
+    train_ds = tfds.as_numpy(ds_builder.as_dataset(split="train", batch_size=-1))
+    test_ds = tfds.as_numpy(ds_builder.as_dataset(split="test", batch_size=-1))
+    train_ds["image"] = jnp.float32(train_ds["image"]) / 255.
+    test_ds["image"] = jnp.float32(test_ds["image"]) / 255.
+    return train_ds, test_ds
+
+
+class MLP(nn.Module):
+    @nn.compact
+    def __call__(self, x):
+        x = nn.relu(nn.Dense(784)(x))
+        x = nn.relu(nn.Dense(200)(x))
+        x = nn.relu(nn.Dense(200)(x))
+        x = nn.Dense(10)(x)
+        return nn.log_softmax(x)
+
+
+@jax.vmap
+def cross_entropy_loss(logits, label):
+    return -logits[label]
+
+
+@jax.jit
+def normal_loss(params, batch):
+    logits = MLP().apply({"params": params}, batch["image"])
+    logits = jax.nn.log_softmax(logits)
+    loss = jnp.mean(cross_entropy_loss(logits, batch["label"]))
+    return loss
+
+
+@jax.jit
+def normal_accuracy(params,batch):
+    logits = MLP().apply({"params": params}, batch["image"])
+    logits = jax.nn.log_softmax(logits)
+    return jnp.mean(jnp.argmax(logits, -1) == batch["label"])
+
+
+@jax.jit
+def theta_to_flat_params(theta,M,flat_params0):
+    return jnp.matmul(theta, M)[0] + flat_params0
+
+
+def projected_loss(theta_subspace, batch, M, flat_params0, reconstruct_fn):
+    """
+    Project the subspace model weights onto the full model weights and
+    compute the loss.
+            w(theta) = w_init + A * theta
```

Comment:
In code we often using longer variables names than in papers, for clarity.
I suggest using `params_subspace` for theta, `params_full` for w, and `params_pytree` for w in tree format.